### DISCUSSION TITLE
Can the Modelica.Blocks.Examples.PID_Controller translate C language and be used as a product environment code?

### DISCUSSION THREAD

**Discussion by pily1** [OpenModelica/OpenModelica#12559] [Q&A] [answered]
Can the Modelica.Blocks.Examples.PID_Controller translate C language and be used as a product environment code?
if it can,how to convert it?
Can the code translated into C language be used as a product environment code when deployed onto a car chip?

**Comment by casella** [comment]
In principle, yes, though you need to generate code in suitable runtime environment. For example, the default C runtime of OMC uses dynamic memory allocation and garbage collection, so it is not suitable for hard real-time; on the other hand, the C++ runtime does not use garbage collection - ABB uses OMC-generated FMUs for industrial predictive control products.

In the past we experimented with small-footprint runtime for Arduino, @sjoelund can comment on that.

**Reply by pily1** [reply]
think you very much casella,If the code conversion from Modelica simulation to C language cannot be practically deployed, what was the original purpose of designing the simulation code to be convertible to C language?

**Reply by casella** [reply]
I'm not sure what you mean by "cannot be practically deployed". OpenModelica can run system simulations by generating C code that is then linked to a runtime. We've been doing that for 20 years. Just not for embedded hardware.

**Reply by pily1** [reply]
I am sorry,what I mean is, after converting a simulation model to C language code, can it be embedded and used within a program written by myself?

**Reply by casella** [reply]
For that, the best option is to export the Modelica code as an [FMU](https://openmodelica.org/doc/OpenModelicaUsersGuide/v1.11.0/fmi.html). Then you can interact with the FMU in C (that's the original choice), or use Python packages such as FMPy or PyFMI

**Reply by pily1** [reply]
What I'm trying to achieve is: Taking the control module as an example, after the simulated model is exported as an FMU, my code's purpose for calling the FMU is not for simulation but to utilize the control code implemented inside the simulation. I want to input a value and precisely get an output value. Can exporting an FMU achieve this effect? If so, how does it convert the non-causal modeling equations into input-output logic?

**Reply by pily1** [reply]
What I'm trying to achieve is: Taking the control module as an example, after the simulated model is exported as an FMU, my code's purpose for calling the FMU is not for simulation but to utilize the control code implemented inside the simulation. I want to input a value and precisely get an output value. Can exporting an FMU achieve this effect? If so, how does it convert the non-causal modeling equations into input-output logic?

**Reply by casella** [reply, marked as answer]
You can generate an FMU-CS (co-simulation) that also embeds the explicit Euler solver for time integration. Then, basically you have functions to set parameters, inputs, and initial states, and functions to advance to the next time step and get the outputs. You can look for any FMI tutorial on the web.

The conversion is done by the OMC backend, in the causalization process. Basically, the equations are ordered automatically to be solved in a sequence, based on their dependencies; if there are coupled equations, they are solved by numerical solvers. I would suggest you the excellent book Cellier Kofman Continuous System Simulation to learn how this works.

**Reply by pily1** [reply]
think you very much,I will try it

**Comment by casella** [comment]
Regarding deployment to car chips, you can follow the development of the [eFMI](https://www.efmi-standard.org/) standard, which was basically born with that goal in mind. OMC currently has no support for that, but it might in the future.

**Reply by pily1** [reply]
think you ,I will try it!

**Reply by sjoelund** [reply]
It is also possible to use the FMI standard. It would then not be for real-time code, but OpenModelica can export it

**Reply by pily1** [reply]
What I'm trying to achieve is: Taking the control module as an example, after the simulated model is exported as an FMU, my code's purpose for calling the FMU is not for simulation but to utilize the control code implemented inside the simulation. I want to input a value and precisely get an output value. Can exporting an FMU achieve this effect? If so, how does it convert the non-causal modeling equations into input-output logic?
